src/cocotb/types/logic_array.py

```diff
@@ -391,15 +391,16 @@ def from_signed(
         """
         if isinstance(range, int):
             range = Range(range - 1, "downto", 0)
-        if value < 0:
-            value += 2 ** len(range)
-        # If value doesn't fit in range, it will still be negative and will blow the
-        # constructor up in a bad way.
-        if value < 0:
+        elif not isinstance(range, Range):
+            raise TypeError("Invalid range type")
```

Comment:
To mirror the check on `__init__`.
```suggestion
            raise TypeError(f"Expected Range or int for parameter 'range', not {type(range).__qualname__}")
```